### DISCUSSION TITLE
Update 2022-09-14

### DISCUSSION THREAD

**Discussion by kaveh808** [kaveh808/kons-9#107] [Announcements]
Some early OpenGL GUI results using the work-in-progress text engine.

message-box:
![Screen Shot 2022-09-14 at 4 36 53 PM](https://user-images.githubusercontent.com/103860793/190301945-505c284f-df58-4f54-b860-3d2cd2786545.png)

dialog-box:
![Screen Shot 2022-09-14 at 4 37 24 PM](https://user-images.githubusercontent.com/103860793/190301991-a5c3d026-62c0-4418-8632-a258ca4b978b.png)

inspector:
![Screen Shot 2022-09-14 at 4 44 49 PM](https://user-images.githubusercontent.com/103860793/190302034-72e8183b-c2b0-4724-8a7b-bb40b8b63538.png)

**Comment by awolven** [comment]
try doing a text-edit box

**Reply by kaveh808** [reply]
It's on my list. Is there a way to get the pixel width of text strings? Or of characters/glyphs?

**Reply by JMC-design** [reply]
I got something that can give you a precise measure, but I'm not sure we need something like that.  Can just divide the space you got by the pixels per em of the font. If it's a monospace font, which most people prefer for code, then all the better as it would be pretty exact.
I had been hoping of just using the pretty printer with miser width but it doesn't seem to be working for me anymore on the latest sbcl.

**Reply by kaveh808** [reply]
For text editing, I plan on using a fixed-width font. For the GUI text I prefer a variable width font, which is where we need the widths to center labels and such.

**Reply by JMC-design** [reply]
This is what I currently use https://plaster.tymoon.eu/view/3423#3423 for determining extents of a string.  It's a bit simpler if you only want width but then you may end up with that horrible UX of chopping of the head and feet of large text.
it doesn't do tracking yet, and it doesn't work for unscaled fonts, I've raised the issue with zpb-ttf.
However, it'll have to be changed to deal with however the fonts are transformed to work with the text engine, unless you keep track of the font somewhere.
sorry for the late reply, I've been a bit busy this week as I've got to make some new dance shoes for saturday.

**Reply by kaveh808** [reply]
Yes, that is the sort of thing I'll need. Width will do for now.

**Comment by jolby** [comment]
This is great!

**Reply by kaveh808** [reply]
Thanks.
